Add tizen armel build CI for pull request

clamp03 opened the pull request:
With ubuntu-18.04-cross-armel-tizen docker image,
check Tizen_armel build for each pull reqest.

clamp03 committed:
Add tizen amrel build CI

clamp03 commented:
Currently, it fails before https://github.com/dotnet/runtime/pull/55789 is merged.

am11 commented on eng/pipelines/coreclr/templates/build-job.yml:
```diff
@@ -141,6 +141,16 @@ jobs:
     - name: SignType
       value: $[ coalesce(variables.OfficialSignType, 'real') ]
 
+    - name: clrRuntimePortableBuildArg
+      value: ''
+    - name: clrManagedPortableBuildArg
+      value: ''
+    - ${{ if eq(parameters.archType, 'armel' )}}:
+      - name: clrRuntimePortableBuildArg
+        value: '-portablebuild=false'
+      - name: clrManagedPortableBuildArg
+        value: '--portablebuild false'
```
Could it be just one variable rather than two? iow, is there a scenario where clrRuntimePortableBuildArg != clrManagedPortableBuildArg? In practical sense, I think it will not work because host resolver that uses runtime graph will actually fail if runtime is -portable and libs are non-portable. The opposite can work, but afaict, it is either PortableBuild=true or false for all subsets alike.

clamp03 replied:
Thanks for the review.
String value of clrRuntimePortableBuildArg and clrManagedPortableBuildArg are different. "-portablebuild=false" vs "--portablebuild false". I can remove clrManagedPortableBuildArg. And then I checked build and execution on a tizen armel device with the instructions in build-job.yml.
Thank you.

am11 replied:
Eventually they all get processed by the name handler: https://github.com/dotnet/runtime/blob/2053d653bfd7fc0a5d073a5d31e01600d2d5fa3f/eng/native/build-commons.sh#L410
In build-runtime, it just strips the extra hyphen, but `-portablebuild=false` should work for clr, libs and host subsets at least.

clamp03 replied:
```build.sh``` uses ```eng/build.sh```.
In ```eng/build.sh``, portable argument is processed not by ```runtime/eng/native/build-commons.sh```
https://github.com/dotnet/runtime/blob/9d9b465f0be92a0d7584c4b8ad6b873be7f0dbc9/eng/build.sh#L403-L414

And I checked ```build.sh``` with ```-portablebuild=false``` too.
It fails to get a DistroRID what I want and it shows below error message.
```
__DistroRid: linux-armel
MSBUILD : error MSB1001: Unknown switch.
Switch: -portablebuild=false

For switch syntax, type "MSBuild -help"
Build failed with exit code 1. Check errors above.
```

am11 replied:
Ah right. Currently approach is to use `/p:PortableBuild=false` with build.sh and build-runtime.sh.

We could actually change https://github.com/dotnet/runtime/blob/9d9b465f0be92a0d7584c4b8ad6b873be7f0dbc9/eng/native/build-commons.sh#L299
to
https://github.com/dotnet/runtime/blob/2053d653bfd7fc0a5d073a5d31e01600d2d5fa3f/eng/build.sh#L166
(with` lowerI` as variable name; strips leading hyphen)

clamp03 replied:
Right, we can change the build script for better argument handling. However, I think it is better to make a new issue (or new PR) for consistent argument handling in build scripts.
Thank you.

am11 commented on eng/pipelines/coreclr/templates/build-job.yml:
```diff
@@ -210,7 +216,7 @@ jobs:
 
     # Build native test components
-    - ${{ if ne(parameters.isOfficialBuild, true) }}:
+    - ${{ if and(ne(parameters.isOfficialBuild, true), ne(parameters.archType, 'armel'))}}:
```
Could it be that it was failing native test subset just because `-portablebuild=false` is not passed?

clamp03 replied:
Right. It needs a portable option and some others.
However, as I know, there is no ARMEL environment for test in the CI.
To prevent build failure, I just want to add Tizen ARMEL build this time.
Thank you.

clamp03 commented:
@safern Could you review this PR? I added Tizen ARMEL Build and it works. However, I think I missed some details. Thank you.

clamp03 commented:
@safern I think errors are not related to this patch. Could you please review it?

safern commented:
> @safern I think errors are not related to this patch. Could you please review it?

Sorry, @clamp03, I was OOF. Will take a look tomorrow.

clamp03 commented:
@safern Thank you! :)

clamp03 committed:
Fix some errors

clamp03 committed:
Fix platform-matrix.yml

clamp03 committed:
Skip tests for tizen-armel

clamp03 committed:
Fix typo

clamp03 commented:
Oops. I pushed the old patch by mistake. Resolved now.

safern commented on eng/pipelines/coreclr/templates/build-job.yml:
```diff
@@ -210,7 +216,7 @@ jobs:
 
     # Build native test components
-    - ${{ if ne(parameters.isOfficialBuild, true) }}:
+    - ${{ if and(ne(parameters.isOfficialBuild, true), ne(parameters.archType, 'armel')) }}:
```
Should we instead introduce a boolean parameter to the template to disable all the test steps rather than conditioning for the archType directly?

clamp03 replied:
Thank you. I updated.

clamp03 replied:
Thank you. I updated.

clamp03 replied:
@safern Is it wrong? Could you please review it? Thank you.

safern replied:
You need to pass down that parameter on `platform-matrix` when in armel and set it to true. That way it will work, at the moment it is just taking the parameter default value to false.

clamp03 replied:
I added ```disableClrTest``` parameter and set it to true in ```platform-matrix```. Is it the right way?

safern commented:
This is currently just adding support for the `coreclr` build job template. Do we want to protect the dotnet/runtime end-to-end build? i.e, building all the subsets and producing a runtime pack for tizen?

clamp03 committed:
Introduce a paramter to skip tests

clamp03 commented:
> This is currently just adding support for the `coreclr` build job template. Do we want to protect the dotnet/runtime end-to-end build? i.e, building all the subsets and producing a runtime pack for tizen?

I think the others are handled by arm32 (hardfp) build. And I have never seen build failure in the other cases.
If armel build fails again, I will update it.
Thank you.

clamp03 commented:
@safern  Could you review it again?

clamp03 commented:
@jkotas @safern Could you give any comments for this PR?

safern commented on eng/pipelines/runtime.yml:
```diff
@@ -72,6 +72,7 @@ jobs:
     - Linux_musl_arm64
     - Linux_musl_x64
     - OSX_arm64
+    - Tizen_armel
```
This will only build coreclr for Tizen, do we want to build other parts of the product as well?

jkotas replied:
It should be enough to just build CoreCLR, for now at least.

The motivation for adding this coverage is to catch Tizen specific build breaks. The Tizen specific build breaks have been pretty much always in CoreCLR.

safern commented:
Sorry @clamp03 I was out for some weeks.

clamp03 commented:
> Sorry @clamp03 I was out for some weeks.

It is Okay. Thank you.

safern reviewed (approved):
LGTM! Thanks 😄

clamp03 commented:
I think failures are not related to this PR. Could you please review and merge?
Thank you.

jkotas commented:
@safern Could you please final look and merge if it looks good?

clamp03 committed:
Update osGroup to Tizen

safern commented:
@clamp03 I see that there is an error on the build coreclr runtime step:

```
/__w/1/s/eng/native/init-distro-rid.sh: line 44: [: missing `]'
/__w/1/s/eng/native/init-distro-rid.sh: line 44: tizen: command not found
```

Also, I see that in that step the distro rid is resolved as: `__DistroRid: tizen.6.5.0-armel` and in `Build managed product components and packages` step it is resolved as: `__DistroRid: linux-armel`

I think we need to fix these?

clamp03 committed:
Fix an error

clamp03 commented:
@safern Thank you for the review
> @clamp03 I see that there is an error on the build coreclr runtime step:
> 
> ```
> /__w/1/s/eng/native/init-distro-rid.sh: line 44: [: missing `]'
> /__w/1/s/eng/native/init-distro-rid.sh: line 44: tizen: command not found
> ```
Sorry. I missed error messages.
This is not related to this Patch. It is because of  https://github.com/dotnet/runtime/pull/59178.
I fixed and added a patch in this PR.

> 
> Also, I see that in that step the distro rid is resolved as: `__DistroRid: tizen.6.5.0-armel` and in `Build managed product components and packages` step it is resolved as: `__DistroRid: linux-armel`
> 
> I think we need to fix this?

In our Tizen, we use clr and libs.native with portable build for tizen and the others for common linux-armel.
So I think we don't need to fix this.

clamp03 committed:
Update fix

clamp03 commented:
An error (multiline comment) in ```runtime (CoreCLR GCC Product BUild Linux x86 Checked)``` is not related to this PR.
It is by https://github.com/dotnet/runtime/pull/58727. There is ```\``` in the end of comment line. I think it is better to fix by the author of PR 58727 in other PR. CC @benjamin-hodgson

@safern @jkotas Could you please review this PR and merge if it is good?

safern commented:
@hoyosjs @jkoritzinsky it seems like the GCC build is broken?
